File: web/tests/reference/graphs/clone-graph.py
def clone_graph(node):
    if node is None:
        return None
    cloned = {}

    def dfs(original):
        if id(original) in cloned:
            return cloned[id(original)]
        copy = Node(original.val)
        cloned[id(original)] = copy
        for neighbor in original.neighbors:
            copy.neighbors.append(dfs(neighbor))
        return copy

    return dfs(node)

File: web/tests/reference/graphs/clone-graph.js
function cloneGraph(node) {
  if (node === null) return null;
  const cloned = new Map();

  const dfs = (original) => {
    if (cloned.has(original)) return cloned.get(original);
    const copy = new Node(original.val);
    cloned.set(original, copy);
    for (const neighbor of original.neighbors) {
      copy.neighbors.push(dfs(neighbor));
    }
    return copy;
  };

  return dfs(node);
}
